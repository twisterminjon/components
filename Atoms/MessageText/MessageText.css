.messagetext {
  display: inline-block;
  font-family: var(--font-message-page--message-bubble-family);
  font-size: var(--font-message-page--message-bubble-size);
  font-weight: var(--font-message-page--message-bubble-weight);

  border-radius: var(--message-border-radius);
  padding: 5px 8px;

  white-space: pre-wrap;
}

.messagetext--sent {
  color: var(--messagepage__sent_fg);
}

.messagetext--received {
  color: var(--messagepage__received_fg);
}

.messagetext-content {
  word-break: break-word;
  margin-bottom: 0;
}

.messagetext-timestamp {
  transition: max-height 0.4s ease-out;
  max-height: 0;
}

.messagetext-timestamp--show {
  max-height: 50px;
}

.messagetext-timestamp-text {
  font-size: 12px;
  color: var(--messagepage__timestamp_fg);
  white-space: nowrap;
  pointer-events: none;
}
