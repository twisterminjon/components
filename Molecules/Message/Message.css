.message {
  display: flex;
  flex: 1;
  flex-direction: column;
  padding: 0 58px 0 14px;
}

.message-content {
  /* display: flex;
  flex-direction: row; */
  border-radius: var(--message-border-radius);
  overflow: hidden;
}

.message--sent {
  background-color: var(--messagepage__sent_bg);
  border: 1px solid var(--messagepage__sent_border);
  border-bottom-right-radius: 0;
}

.message--received {
  background-color: var(--messagepage__received_bg);
  border: 1px solid var(--messagepage__received_border);
  border-bottom-left-radius: 0;
}

.message--received .message-file {
  background-color: var(--messagepage__received_file_bg);
}

.message-flex-container {
  display: flex;
  flex-direction: row;
}

.message-flex-container-column {
  display: flex;
  flex-direction: column;
}

.message--own {
  align-items: flex-end;
  padding: 0 10px 0 106px;
}

.message-author {
  color: var(--font_color_light);
  font-size: var(--font-message-page--author-size);
  font-weight: var(--font-base--bold);
  letter-spacing: -0.34px;
  padding: 3px 8px;
}

.message-avatar {
  display: flex;
}

.message-retry {
  max-height: 0;
  opacity: 0;
  transition: all 200ms ease-out;
}

.message-retry-visible {
  padding-top: 5px;
  max-height: 40px;
  opacity: 1;
}

.message-retry-button {
  background-color: transparent;
  color: var(--button_negative_bg);
  border-radius: 8px;
  border-top-right-radius: 0;
  padding: 7px;
  border: 1px solid var(--button_negative_bg);
  outline: none;
}

.message-retry-button:active {
  transform: scale(1.02);
}
.message-retry-button:hover {
  transform: scale(1.02);
}
