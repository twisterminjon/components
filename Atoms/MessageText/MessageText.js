import React, { useState, useMemo, useCallback } from 'react'
import PropTypes from 'prop-types'

import { DateFormat, ProjectDate } from '@shared/helpers'

import Linkify from '../Linkify/Linkify'

import './MessageText.css'

MessageText.propTypes = {
  /** The type of message bubble to display */
  isOwn: PropTypes.bool,
  /** Timestamp for the message */
  timestamp: PropTypes.string.isRequired,
  /** Content for the text message */
  children: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.node), PropTypes.node]).isRequired,
  /** ClassName for the wrapper */
  className: PropTypes.string,
  /** Style for the wrapper */
  style: PropTypes.object,
}

MessageText.defaultProps = {
  isOwn: false,
  className: '',
  style: {},
}

function MessageText({ isOwn, timestamp, children, className, style, ...rest }) {
  const [showTime, setShowTime] = useState(false)
  const toggleTime = useCallback(() => setShowTime(isShown => !isShown), [])
  const niceTime = useMemo(() => ProjectDate(timestamp).formatLocalUTC(DateFormat.LT), [timestamp])

  const type = isOwn ? 'sent' : 'received'
  const showTimestampClass = showTime ? 'messagetext-timestamp--show' : ''

  return (
    <div onClick={toggleTime} className={`messagetext messagetext--${type} ${className}`} style={style} {...rest}>
      <p className="messagetext-content">
        <Linkify text={children} />
      </p>
      <div className={`messagetext-timestamp ${showTimestampClass}`.trim()}>
        <span className="messagetext-timestamp-text">{niceTime}</span>
      </div>
    </div>
  )
}

export default MessageText
