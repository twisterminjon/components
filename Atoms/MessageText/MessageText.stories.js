import React from 'react'

import { storiesOf } from '@storybook/react'
import centered from '@storybook/addon-centered/react'
import { withKnobs, text, boolean } from '@storybook/addon-knobs'

import MessageText from './MessageText'

const stories = storiesOf('Provider/v1/Atoms', module)
stories.addDecorator(withKnobs)

stories.addDecorator(centered).add('MessageText', () => {
  const isOwnLab = 'isOwn'
  const isOwnDef = false
  const isOwnVal = boolean(isOwnLab, isOwnDef)

  const messageLab = 'message'
  const messageDef =
    'Lorem ipsum dolor sit amet consectetur adipisicing elit. Repudiandae facere dolores vero. Eum laborum dicta nesciunt eveniet exercitationem! Odio esse recusandae voluptate totam pariatur quis eum amet soluta ipsam distinctio.'
  const messageVal = text(messageLab, messageDef)

  const timestampLab = 'timestamp'
  const timestampDef = '2020-08-08T13:30:00'
  const timestampVal = text(timestampLab, timestampDef)

  return (
    <div
      style={{
        width: 400,
      }}>
      <MessageText timestamp={timestampVal} isOwn={isOwnVal}>
        {messageVal}
      </MessageText>
    </div>
  )
})
