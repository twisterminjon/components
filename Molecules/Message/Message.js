import React from 'react'
import PropTypes from 'prop-types'

import { MessageStatus } from '@shared/helpers'

import Avatar from '../../Atoms/Avatar/Avatar'

import './Message.css'

Message.propTypes = {
  /** Can have a different appearance if message belongs to sender */
  isOwn: PropTypes.bool,
  /** Timestamp for the message */
  timestamp: PropTypes.string.isRequired,
  /** Name of the user for the message */
  displayName: PropTypes.string.isRequired,
  /** Path to the users profile pic */
  profileImage: PropTypes.string,
  /** Status of message */
  status: PropTypes.oneOf(Object.values(MessageStatus)),
  /** Called when message is failed and retry pressed */
  onRetry: PropTypes.func.isRequired,
  /** Content for the text message */
  children: PropTypes.oneOfType([PropTypes.node]).isRequired,
  /** ClassName for the wrapper */
  className: PropTypes.string,
  /** Style for the wrapper */
  style: PropTypes.object,
}

Message.defaultProps = {
  isOwn: false,
  profileImage: '',
  className: '',
  style: {},
}

const FailedStatus = [MessageStatus.Failed, MessageStatus.Retrying]

export default function Message({
  isOwn,
  timestamp,
  displayName,
  profileImage,
  status,
  onRetry,
  className,
  style,
  children,
  ...rest
}) {
  // clone the child and add isOwn prop
  const clonedChild = React.cloneElement(children, { isOwn, timestamp })
  const typeClass = isOwn ? 'message--own' : ''
  const messageClass = isOwn ? 'message--sent' : 'message--received'
  const messageFileClass = children.type.name === 'MessageAttachmentFile' ? 'message-file' : ''
  const retryClass = FailedStatus.includes(status) ? 'message-retry-visible' : ''

  return (
    <div className={`message ${typeClass} ${className}`.trim()} style={style} {...rest}>
      <div className="message-flex-container">
        <div className={`message-content ${messageClass}`.trim()}>
          {!isOwn && (
            <div className={`message-avatar ${messageFileClass}`.trim()}>
              <Avatar size={24} imgUrl={profileImage} displayName={displayName} />
              <span className="message-author">{displayName}</span>
            </div>
          )}
          <div className="message-flex-container-column">{clonedChild}</div>
        </div>
      </div>
      <div className={`message-retry ${retryClass}`.trim()}>
        <button className="message-retry-button" disabled={MessageStatus.Retrying === status} onClick={onRetry}>
          {MessageStatus.Retrying === status ? 'Retrying...' : 'Send Failed. Retry?'}
        </button>
      </div>
    </div>
  )
}
