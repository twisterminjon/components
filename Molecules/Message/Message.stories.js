import React from 'react'

import { storiesOf } from '@storybook/react'
import centered from '@storybook/addon-centered/react'
import { withKnobs, boolean, select } from '@storybook/addon-knobs'
import { action } from '@storybook/addon-actions'

import { MessageStatus } from '@shared/helpers'

import Message from './Message'
import MessageText from '../../Atoms/MessageText/MessageText'

const stories = storiesOf('Provider/v1/Molecules', module)
stories.addDecorator(withKnobs)

stories.addDecorator(centered).add('Message', () => {
  const textLong =
    'Lorem ipsum dolor sit amet consectetur adipisicing elit. Repudiandae facere dolores vero. Eum laborum dicta nesciunt eveniet exercitationem! Odio esse recusandae voluptate totam pariatur quis eum amet soluta ipsam distinctio.'
  const textShort = 'What up, dog?'

  const textLab = 'Short Text'
  const textDef = false
  const textVal = boolean(textLab, textDef)
  const textDisplay = textVal ? textShort : textLong

  const isOwnLab = 'isOwn'
  const isOwnDef = true
  const isOwnVal = boolean(isOwnLab, isOwnDef)

  const status = select('status', MessageStatus, MessageStatus.Succeed)

  return (
    <div
      style={{
        position: 'absolute',
        right: 0,
        left: 0,
        top: 0,
        bottom: 0,
        backgroundColor: 'rgba(255,255,255,0.05)',
      }}>
      <Message
        status={status}
        isOwn={isOwnVal}
        onRetry={action('onRetry')}
        profileImage="https://www.fillmurray.com/60/60"
        timestamp="2018-12-05T10:30:15">
        <MessageText>{textDisplay}</MessageText>
      </Message>
    </div>
  )
})
