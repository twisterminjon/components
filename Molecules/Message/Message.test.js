import React from 'react'
import renderer from 'react-test-renderer'

import Message from './Message'
import MessageText from '../../Atoms/MessageText/MessageText'

describe('Message', () => {
  it('matches the snapshot', () => {
    const comp = renderer.create(
      <Message
        isOwn={false}
        onRetry={() => undefined}
        timestamp="2020-08-08T13:30:15"
        profileImage=""
        displayName="test">
        <MessageText timestamp="2020-08-08T13:30:15">test message</MessageText>
      </Message>
    )

    expect(comp.toJSON()).toMatchSnapshot()
  })
})
