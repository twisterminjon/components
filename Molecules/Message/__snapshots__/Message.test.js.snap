// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Message matches the snapshot 1`] = `
<div
  className="message"
  style={Object {}}
>
  <div
    className="message-flex-container"
  >
    <div
      className="message-content message--received"
    >
      <div
        className="message-avatar"
      >
        <div
          className="avatar"
          style={
            Object {
              "height": "24px",
              "lineHeight": "24px",
              "width": "24px",
            }
          }
        >
          <span>
            t
          </span>
        </div>
        <span
          className="message-author"
        >
          test
        </span>
      </div>
      <div
        className="message-flex-container-column"
      >
        <div
          className="messagetext messagetext--received "
          onClick={[Function]}
          style={Object {}}
        >
          <p
            className="messagetext-content"
          >
            test message
          </p>
          <div
            className="messagetext-timestamp"
          >
            <span
              className="messagetext-timestamp-text"
            >
              1:30 PM
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="message-retry"
  >
    <button
      className="message-retry-button"
      disabled={false}
      onClick={[Function]}
    >
      Send Failed. Retry?
    </button>
  </div>
</div>
`;
